Implement react-intl for Welsh language support

aliblackwell opened the pull request:
This PR adds a toggle in the footer for English/Welsh. Current translations are from a free online translations service and will need to be done properly.

Details here: https://www.codeandweb.com/babeledit/tutorials/how-to-translate-your-react-app-with-react-intl

The new structure demands a comprehensive refactor of our test suite to use react-testing-library as any components that were previously shallow rendered now require the withTranslations.js parent component to render properly. 

Next steps are to get it all working with Babel Edit.

chris48s commented:
I've not had a properly detailed look at this yet, but there are quite a lot of strings that aren't pulled out for translation yet. This probably isn't exhaustive, but I noticed:

* We don't know of any upcoming elections in your area
* Back to Postcode search
* Your polling station
* Show me directions
* We couldn't find your station
* Get in touch with
* Phone
* Website
* Email

There are probably more. As the next thing, can you go through and find all the remaining hard-coded strings and pull them out for translation. You can just shove the Welsh versions in as a placeholder e.g: "XXX" for the moment. We'll send them all off to a translator anyway. If we can get to the stage where we've isolated all the strings, then we can kick that process off. Cheers

aliblackwell commented:
Good shout – this is exactly what I wanted to spend more time doing. And actually, the "my address is not in this list" option was missing in this PR! Somehow I missed that string and as you point out, the test suite didn't catch it. I let myself become beguiled by the coveralls percentage creeping ever-upwards. I am now working through all the tests that got cut in the refactor, re-implementing them using the new format. We will have a much more comprehensive, robust and far less brittle test suite at the end of this process.

aliblackwell commented:
Below is an audit of all the tests present at commit a812660556d5a1da176ed522dd1a0f0054b9442a on April 22nd this year.

Old test: StationNotFound.test.js 'shows notification when there is a voter id pilot'
Replaced by: Notifications.test.js 'should show voter ID requirement for DE13GB'

Old test: Widget.test.js 'should declare that it cannot find the station if so' 
Replaced by: StationNotFound.test.js 'should show a station not found page for unknown polling stations'

Old test: StationNotFound.test.js: ''should present contact details' 'via Phone' 'via Email 'via Website'
Replaced by: StationNotFound.test.js "should present council to get in touch with when station isn't found" – multiple assertions in single test

Old test: StationNotFound.test.js "does not show notification when there is no voter id pilot'"
Replaced by: Notification.test.js: 'does not show notification when there is no voter id pilot'

Old test: StationNotFound.test.js "shows notification when there is a voter id pilot'
Replaced by: Notification.test.js: 'should show voter ID requirement for DE13GB'

Old test: ResultsCard.test.js "renders header"
Replaced by: PollingStation.test.js 'shows a title if it finds a polling station'

Old test: ResultsCard.test.js "renders address"
Replaced by: PollingStation.tes.js "renders address"

Old test: ResultsCard.test.js 'does not show notification when there is no voter id pilot'
Replaced by: Notifications.test.js 'does not show notification when there is no voter id pilot'

Old test: ResultsCard.test.js 'shows notification when there is a voter id pilot' (duplicate of StationNotFound.test.js above)
Replaced by: Notification.test.js: 'should show voter ID requirement for DE13GB'

**Action required**

Old test: StationNotFound.test.js: 'should present NI Electoral Office for N09 postcodes'
Not replaced. It would be useful to have some example JSON for a response that requires this message.

Old test: StationNotFound.test.js: 'always renders header'
Replaced by: DemocracyClubWidget.test.js: 'should include an h1 at the top level of the postcode selector' But needs implementing for StationNotFound view.

N.B. The tests for analytics / referrer code haven't been touched since I took on this project and it would be useful to discuss the requirements for these to ensure the widget still meets them.

chris48s commented:
Right, just to summarise the points from the call about testing

* Lets not add extraneous example responses if we can avoid it. For example, there are several example responses added in this PR which don't get referenced in any test case. For example:
  * AA13AA
  * AA14AA
  * AA15AA
  * AL71RG
  These can all be removed.
* If there are components which can be tested in isolation, its useful to unit test them to keep the test data/cases small and easy to understand.
* If there are components which we can't test in isolation because of the way we're doing the translations, using a full example response is a totally valid tradeoff.
* If we don't need to create a new example response, use an existing one if possible, or we could slightly modify an existing one (for example if we want to test what happens if we've got a NULL station point and all our existing responses have a station point, we could modify one of them).
* If we're going to create an example response re-creating an particular condition, the example response should be something that the upstream API could plausibly serve up. Feel free to ask questions if you're unsure on that, but:
  * If we aren't able to geocode the user's location, the `postcode_location` key will exist, but it will be `null` e.g: `"postcode_location": null`, but the key will always exist in the response.
  * If we've got a polling station with an address but no co-ordinates, that will be represented as a feature with a `null` `geometry` e.g: 

  ```json
  "station": {
    "id": "W06000011.84-st-helens-primary-school",
    "type": "Feature",
    "geometry": null,
    "properties": {
      "postcode": "SA1 3TY",
      "address": "St. Helens Primary School\nVincent Street\nSwansea"
    }
  }
  ```
* If we do need to add new example response to recreate conditions for testing purposes, can we document them in `000_README.md` so we know what its for.

chris48s commented on public/index.html:
```diff
@@ -95,7 +95,7 @@ <h1>Lorem Ipsum Town Council</h1>
             scrolling="no"></iframe>
         </noscript>
-        <div id="dc_wdiv" aria-live="polite" role="region"></div>
+        <div id="dc_wdiv" data-language="en" data-language-toggle="true" aria-live="polite" role="region"></div>
```
This isn't a _huge_ deal, but do you reckon we definitely need to make this 2 settings? I reckon we could squash it to:

`<div id="dc_wdiv">` - don't show the language picker at all
`<div id="dc_wdiv" data-language="en">` - show the language picker and default it to English
`<div id="dc_wdiv" data-language="cy">` - show the language picker and default it to Welsh

aliblackwell replied:
Are we sure there aren't going to be people who want it only in Welsh? As if not, this is simpler.

aliblackwell replied:
This is now implemented.

aliblackwell replied:
https://github.com/DemocracyClub/WhereDoIVote-Widget/pull/281/commits/835fa85f9714c2f05e866b921be21cd8778fefa5

chris48s commented on src/Branding.js:
```diff
@@ -8,43 +9,54 @@ function ErrorMessage(props) {
-        Built by{' '}
-        <img alt="Democracy Club" src="https://widget.wheredoivote.co.uk/logo-with-text.png" />
+        {formatMessage({ id: 'general.built-by' })}
+        <img alt={formatMessage({ id: 'general.dc-club' })} src="/img/logo-with-text.png" />
```
The reason this was a hard-coded URL is because once we're embedded in someone else's site our base URL for relative URLs becomes `manchestereveningnews.co.uk` or `cardiff.gov.uk` or whatever, not `widget.wheredoivote.co.uk` so it won't work as a relative URL.

As a short term fix, can we just change it back to a hard-coded URL.

Bit of a tangent - lets not do it in this PR - but one thing I've thought about is switching out this PNG for an inline SVG - then we can just compress the SVG into the minified bundle. My reservation is if we have to pull in an extra react plugin to make a component that renders an inline SVG, it might just be counter-productive. Its not really the important thing here anyway..

aliblackwell replied:
Have reverted this – good point about the relative URLs. 

The SVG should just work without the need for any fancy-pants plugins. But yep, let's do that in a future PR.

chris48s commented:
> Old test: StationNotFound.test.js: 'should present NI Electoral Office for N09 postcodes'
> Not replaced. It would be useful to have some example JSON for a response that requires this message.

https://developers.democracyclub.org.uk/api/v1/sandbox/postcode/AA15AA/ will do it, but if we can just do it as a unit test at the component level, that would be quite nice.

> Old test: StationNotFound.test.js: 'always renders header'
> Replaced by: DemocracyClubWidget.test.js: 'should include an h1 at the top level of the postcode selector' But needs implementing for StationNotFound view.

Not sure I understand what the question/blocker is on this one?

> N.B. The tests for analytics / referrer code haven't been touched since I took on this project and it would be useful to discuss the requirements for these to ensure the widget still meets them.

All looks fine to me :+1:

chris48s commented:
Oh, just so all the feedback is in one place..

I mentioned this morning about the `[React Intl] Missing locale data for locale: "cy". Using default locale: "en" as fallback.` warnings when we run the tests in CI (and locally for me). If you could work out what's up there, that would be great too.

aliblackwell commented:
@chris48s this is good to go!

chris48s commented on src/DCWidgetEdgeCases.test.js:
```diff
@@ -1,111 +1,47 @@
+    const YourPollingStation = await waitForElement(() =>
+      document.querySelector('.PollingStation')
+    );
+    expect(YourPollingStation).toHaveTextContent(en_messages['station.your-station']);
```
I'm confused by this test.
The description seems to imply we're supposed to be testing that we only render the polling station details once (even if there are multiple ballots), but then the actual implementation is just checking that the polling station details are somewhere in the response (which would be true even if we had rendered it multiple times).
What is this actually doing/supposed to be doing?

aliblackwell replied:
Yes good spot, I messed this one up. I have removed it. I'd rather make a list of tests we want to implement in future. Given this has nothing to do with translations I am not going to implement this one now.

aliblackwell replied:
This test has been re-implemented here: 87c8780c61b46bc6ccd55701d85bb0ce5600e9a5

chris48s commented on src/DemocracyClubWidget.test.js:
```diff
@@ -1,48 +1,110 @@
+    submitPostcode();
+    const errorResponse = await waitForElement(() => document.querySelector('.ErrorMessage'));
+
+    expect(errorResponse).toHaveTextContent(en_messages['api.errors.generic-error']);
```
This isn't a very useful test because you're making an assertion about the behaviour of the mock response server you've set up, not testing how the application performs in real life. Given this input, the real API would return a `400 Bad Request` with body `{"message": "Could not geocode from any source" }` and then the error would be `api.errors.bad-postcode`. If we're going to have a test for this case, we should actually replicate the behaviour of the live API in the test. If its a PITA, we can remove this test and punt this to an issue.

aliblackwell replied:
It's using the mock server as a means to an end which is testing that an error is displayed if a bad postcode is entered. I'm happy to delete it.

aliblackwell replied:
But I do see the futility of it. We have postcode tests in PostcodeSelector.test.js. These could probably be fleshed out but I don't think in this PR.
